### DISCUSSION TITLE
Local undo/redo (e.g., paragraph based)

### DISCUSSION THREAD

**Discussion by eugenesvk** [helix-editor/helix#1588] [Ideas]
Let's say I have two separate chunks of text, but within one file (e.g., two paragraphs or two functions)
I make 5 changes to the first one
Then I make 3 changes to the second one
Now my undo history stack is 5 changes from the first and 3 changes from the second and to get to the changes in the first chunk/function I'd need to undo also the 3 changes from the second

Wouldn't it be great if I were able instead to **select the first function** and then execute an undo command, but then only use the **first 5 changes** that are **relevant** to this selected chunk of text?

It's similar to how you can currently undo changes in one file without affecting the later changes in a different file, but for local chunks of text within one file

Or maybe not and that's something that should be left to the likes of git where you can view and undo changes by line/hunk

**Comment by dead10ck** [comment]
>I make 5 changes to the first one
Then I make 3 changes to the second one
Now my undo history stack is 5 changes from the first and 3 changes from the second and to get to the changes in the first chunk/function I'd need to undo also the 3 changes from the second

Typically in the modal editing paradigm, this would involve:

1. In normal mode, navigate to the place where you edit your text
2. Insert stuff by going into insert mode, or delete stuff with other commands like delete.
3. Go back to normal mode, navigate to the second function
4. Make your changes there
5. Go back to normal mode

This way, pressing undo once would undo your changes from the second function, and pressing again would undo the changes from the first.

Navigation also typically takes place with hjkl, so your fingers don't leave the home row. So often times the undo history is even more granular than all the changes you've made to each function as a whole. Your undo history can be as granular as you want it to be.

Helix is a modal editor. It seems like you are trying to make it work as a non-modal editor, and it's not working out for you. If this paradigm is too cumbersome for you, then perhaps you would be happier with a different editor.

**Reply by eugenesvk** [reply]
> Your undo history can be as granular as you want it to be.

Then why did you limit my example to only having 1+1 undos instead 5+3???

But even in your even more trivial example — why would you want to __lose__ the changes in the __second__ function if you want to __undo__ changes in the __first__? Unless the undo stack if intertwined (i.e. you have changes that affect both functions), these could be cleanly separated, so you would ideally be able to limit your undos to the stack pertaining only to a particular function
Just like a git client allows you to review/discard (but not redo) changes on a line/block basic, so could a text editor
Or just like a text editor allows you to do when we consider separate files
Or just like Emacs allows you to do since '98 — you mark the text containing the function and your regular undo only affects this selection, so your changes to the second function aren't undone

And modality has nothing to do with that, not sure what you've confused it with, this applies to modal and non-modal editing alike

**Reply by pickfire** [reply]
If any this would be a plugin I guess given its niche, at least I don't find myself needing this yet.

**Comment by sudormrfbin** [comment]
Similar functionality in emacs: https://www.emacswiki.org/emacs/SelectiveUndo
